Failing to build on certain embedded platforms

Originally reported by @ErickTorresBrownU in https://github.com/rust-random/rand/issues/1574#issuecomment-2687154986:

> For instance, I can no longer compile my embedded Rust code because rand/rand_core now depends on zerocopy.
>
> ```text
> error[E0080]: evaluation of constant value failed
>    --> C:\Users\erick\.cargo\registry\src\index.crates.io-6f17d22bba15001f\zerocopy-0.8.20\src\layout.rs:120:80
>     |
> 120 |     pub(crate) const CURRENT_MAX_ALIGN: NonZeroUsize = match NonZeroUsize::new(1 << 28) {
>     |                                                                                ^^^^^^^ attempt to shift left by `28_i32`, which would overflow
> ```


@ErickTorresBrownU, what platform are you building for here? We would like to support this use case and add this platform to our CI so we don't regress.

Okay thanks. What build command are you using? I'm trying but getting this failure:

```text
$ RUSTFLAGS='-C target-cpu=atmega328p' cargo +nightly-2025-02-25 check --target=avr-none -Zbuild-std
   Compiling gimli v0.31.1
   Compiling miniz_oxide v0.8.3
error[E0432]: unresolved import `alloc::sync`
 --> .../.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/gimli-0.31.1/src/read/dwarf.rs:2:12
  |
2 | use alloc::sync::Arc;
  |            ^^^^ could not find `sync` in `alloc`

error[E0432]: unresolved import `alloc::sync`
 --> .../.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/gimli-0.31.1/src/read/abbrev.rs:4:12
  |
4 | use alloc::sync::Arc;
  |            ^^^^ could not find `sync` in `alloc`

error: values of the type `[u8; 32768]` are too big for the target architecture
  --> .../.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/miniz_oxide-0.8.3/src/inflate/stream.rs:86:5
   |
86 |     fn default() -> Self {
   |     ^^^^^^^^^^^^^^^^^^^^

error: could not compile `miniz_oxide` (lib) due to 1 previous error
warning: build failed, waiting for other jobs to finish...
For more information about this error, try `rustc --explain E0432`.
error: could not compile `gimli` (lib) due to 2 previous errors
```

By contrast, if I build without `-Zbuild-std`, I get:

```text
$ RUSTFLAGS='-C target-cpu=atmega328p' cargo +nightly-2025-02-25 check --target=avr-none
    Checking zerocopy v0.8.20 (.../workspace/zerocopy)
error[E0463]: can't find crate for `core`
  |
  = note: the `avr-none` target may not be installed
  = help: consider downloading the target with `rustup target add avr-none`
  = help: consider building the standard library from source with `cargo build -Zbuild-std`

For more information about this error, try `rustc --explain E0463`.
error: could not compile `zerocopy` (lib) due to 1 previous error
```

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -631,6 +631,45 @@ jobs:
       - name: Check big endian for aarch64_be-unknown-linux-gnu target
         run: ./cargo.sh +nightly build --target=aarch64_be-unknown-linux-gnu -Zbuild-std --features simd
 
+  # We can't use this as part of the build matrix because rustup doesn't support
+  # the `avr-none` target.
+  check_avr_artmega:
+    runs-on: ubuntu-latest
+    name: Build (zerocopy / nightly / --simd / avr-none)
+    steps:
+      - uses: actions/checkout@eef61447b9ff4aafe5dcd4e0bbf5d482be7e7871 # v4.2.1
+      - name: Configure environment variables
+        run: |
+          set -eo pipefail
+          ZC_TOOLCHAIN="$(./cargo.sh --version nightly)"
+          RUSTFLAGS="$RUSTFLAGS $ZC_NIGHTLY_RUSTFLAGS"
+          echo "RUSTFLAGS=$RUSTFLAGS" >> $GITHUB_ENV
+          echo "ZC_TOOLCHAIN=$ZC_TOOLCHAIN" >> $GITHUB_ENV
+      - name: Rust Cache
+        uses: Swatinem/rust-cache@23bce251a8cd2ffc3c1075eaa2367cf899916d84 # v2.7.3
+        with:
+          key: avr-none
+      - name: Install stable Rust for use in 'cargo.sh'
+        uses: dtolnay/rust-toolchain@00b49be78f40fba4e87296b2ead62868750bdd83 # stable
+        with:
+          toolchain: stable
+      - name: Install Rust with nightly toolchain (${{ env.ZC_TOOLCHAIN }})
+        uses: dtolnay/rust-toolchain@00b49be78f40fba4e87296b2ead62868750bdd83 # stable
+        with:
+            toolchain: ${{ env.ZC_TOOLCHAIN }}
+            components: clippy, rust-src
+      # NOTE: We cannot check tests because of a number of different issues (at
+      # the time of writing):
+      # - No `alloc::sync`
+      # - Values of type `[u8; 32768]` are too big
+      #
+      # To try for yourself, replace `-Zbuild-std=core` with `-Zbuild-std` and
+      # add `--tests`.
+      - name: Check avr-none target
+        run: RUSTFLAGS='-C target-cpu=atmega328p' ./cargo.sh +nightly check --target=avr-none -Zbuild-std=core --features simd,simd-nightly,float-nightly,derive
+      - name: Clippy check avr-none target
+        run: RUSTFLAGS='-C target-cpu=atmega328p' ./cargo.sh +nightly clippy --target=avr-none -Zbuild-std=core --features simd,simd-nightly,float-nightly,derive
+
   check_fmt:
     runs-on: ubuntu-latest
     name: Check Rust formatting
@@ -761,7 +800,7 @@ jobs:
       # https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/collaborating-on-repositories-with-code-quality-features/troubleshooting-required-status-checks#handling-skipped-but-required-checks
       if: failure()
       runs-on: ubuntu-latest
-      needs: [build_test, kani,check_be_aarch64 , check_fmt, check_readme, check_versions, generate_cache, check-all-toolchains-tested, check-job-dependencies, run-git-hooks]
+      needs: [build_test, kani, check_be_aarch64, check_avr_artmega, check_fmt, check_readme, check_versions, generate_cache, check-all-toolchains-tested, check-job-dependencies, run-git-hooks]
       steps:
         - name: Mark the job as failed
           run: exit 1

--- a/src/util/macro_util.rs
+++ b/src/util/macro_util.rs
@@ -25,6 +25,7 @@ use core::{
 // TODO(#29), TODO(https://github.com/rust-lang/rust/issues/69835): Remove this
 // `cfg` when `size_of_val_raw` is stabilized.
 #[cfg(__ZEROCOPY_INTERNAL_USE_ONLY_NIGHTLY_FEATURES_IN_TESTS)]
+#[cfg(not(target_pointer_width = "16"))]
 use core::ptr;
 
 use crate::{
@@ -105,11 +106,13 @@ impl<T, U> MaxAlignsOf<T, U> {
 }
 
 #[cfg(__ZEROCOPY_INTERNAL_USE_ONLY_NIGHTLY_FEATURES_IN_TESTS)]
+#[cfg(not(target_pointer_width = "16"))]
 const _64K: usize = 1 << 16;
 
 // TODO(#29), TODO(https://github.com/rust-lang/rust/issues/69835): Remove this
 // `cfg` when `size_of_val_raw` is stabilized.
 #[cfg(__ZEROCOPY_INTERNAL_USE_ONLY_NIGHTLY_FEATURES_IN_TESTS)]
+#[cfg(not(target_pointer_width = "16"))]
 #[repr(C, align(65536))]
 struct Aligned64kAllocation([u8; _64K]);
 
@@ -122,6 +125,7 @@ struct Aligned64kAllocation([u8; _64K]);
 // TODO(#29), TODO(https://github.com/rust-lang/rust/issues/69835): Remove this
 // `cfg` when `size_of_val_raw` is stabilized.
 #[cfg(__ZEROCOPY_INTERNAL_USE_ONLY_NIGHTLY_FEATURES_IN_TESTS)]
+#[cfg(not(target_pointer_width = "16"))]
 pub const ALIGNED_64K_ALLOCATION: NonNull<[u8]> = {
     const REF: &Aligned64kAllocation = &Aligned64kAllocation([0; _64K]);
     let ptr: *const Aligned64kAllocation = REF;

